app/controllers/queue_items_controller.rb

```diff
@@ -14,11 +14,37 @@ def create
   def destroy
     queue_item = QueueItem.find(params[:id])
     queue_item.delete if current_user.queue_items.include?(queue_item)
+    normalize_queue_item_positions
+    redirect_to my_queue_path
+  end
+
+  def update_queue
+    begin 
+      update_queue_items
+      normalize_queue_item_positions
+    rescue ActiveRecord::RecordInvalid
+      flash[:error] = "invalid position number"
+    end
     redirect_to my_queue_path
   end
 
   private
 
+  def update_queue_items
+    ActiveRecord::Base.transaction do 
+      params[:queue_items].each do |queue_item_data|
+      queue_item = QueueItem.find(queue_item_data["id"])
+      queue_item.update_attributes!(position: queue_item_data["position"], rating: queue_item_data["rating"]) if queue_item.user == current_user
+      end
+    end
+  end
+
+  def normalize_queue_item_positions
```

Comment:
This is that kind of behavior that could be moved to User class. You are working with current_user items anyway. I would recommend moving this to User model - also because you will be able to use such functionality in a different controller or model or view. Everytime you try to refactor think about it this way - is this behavior specific to some object ? Is it connected to some model?